docs/QuickStart.md

```diff
@@ -0,0 +1,204 @@
+        spider::Future<std:tuple<int, int>> future = context.run(gcd_impl, x, y);
+        x = future.get().get().get<0>();
+        y = future.get().get().get<1>();
+    }
+    return x;
+}
+
+auto gcd_impl(spider::Context& context, int x, int y) -> std::tuple<int, int> {
+    return { x, x % y};
+}
+```
+
+## Data on External Storage
+Often simple POD data are not enough. However, passing large amount of data around is expensive. Usually these data is stored on disk or a distributed storage system. For example, an ETL workload usually reads in data from an external storage, writes temporary data on an external storage, and writes final data into an external storage.
+
+Spider lets user pass the metadata of these data around in `spider::Data` objects. `Data` stores the value of the metadata information of external data, and provides crucial information to Spider for correct and efficient scheduling and failure recovery. `Data` stores a list of nodes which has locality of the external data, and user can specify if locality is a hard requirement, i.e. task can only run on the nodes in locality list. `Data` can include a `cleanup`function, which will run when the `Data` object is no longer reference by any task and client. `Data` has a persist flag to represent that external data is persisted and do not need to be cleaned up.
+
+```c++
+struct HdfsFile {
+    std::string url;
+};
+
+auto filter(spider::Data<Hdfsfile> input) -> spider::Data<HdfsFile> {
+    std::string const output_path = std::format("/path/%s", context.task_id());
+    std::string const input_path = input.get().url;
+    // Create HdfsFile Data first in case task fails and Spider can clean up the data.
+    spider::Data<HdfsFile> output = spider::Data<HdfsFile>::Builder()
+        .cleanup([](HdfsFile const& file) { delete_hdfs_file(file); })
+        .build(HdfsFile { output_path });
+    auto file = hdfs_create(output_path);
+    std::vector<std::string> nodes = hdfs_get_nodes(file);
+    output.set_locality(nodes, false); // not hard locality
+    
+    run_filter(input_path, file);
+    
+    return output;
+}
+
+auto map(spider::Data<HdfsFile> input) -> spider::Data<HdfsFile> {
+    std::string const output_path = "/path/to/output";
+    std::string const input_path = input.get().url;
+    
+    spider::Data<HdfsFile> output = spider::Data<HdfsFile>::Builder()
+        .cleanup([](HdfaFile const& file) { delete_hdfs_file(file); })
+        .build(HdfsFile { output_path });
+    
+    run_map(input_path, output_path);
+    
+    // Now that map finishes, the file is persisted on Hdfs as output of job.
+    output.mark_persist();
+    return output;
+}
+
+auto main(int argc, char** argv) -> int {
+    spider::Data<HdfsFile> input = spider::Data<HdfsFile>::Builder()
+        .mark_persist(true)
+        .build(HdfsFile { "/path/to/input" });
+    spider::Future<spider::Data<HdfsFile>> future = spider::run(
+        spider::bind(map, filter),
+        input);
```

Comment:
`input` is not a POD, so that violates the constraint mentioned above about task arguments being POD.